File: applications/holoscan_ros2/adi/adi_imu/cpp/adi_imu_ros2_subscriber.cpp
#include <cstdio>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/temperature.hpp>

class ImuRos2Subscriber : public rclcpp::Node {
public:
  ImuRos2Subscriber() : Node("adi_imu_ros2_subscriber") {
    imu_subscription_ = create_subscription<sensor_msgs::msg::Imu>(
        "imu/data_raw", 10,
        [this](sensor_msgs::msg::Imu::ConstSharedPtr message) {
          RCLCPP_INFO(get_logger(), "IMU gyro=(%.4f, %.4f, %.4f) accel=(%.4f, %.4f, %.4f)",
                      message->angular_velocity.x, message->angular_velocity.y,
                      message->angular_velocity.z, message->linear_acceleration.x,
                      message->linear_acceleration.y, message->linear_acceleration.z);
        });
    temperature_subscription_ = create_subscription<sensor_msgs::msg::Temperature>(
        "imu/temperature", 10,
        [this](sensor_msgs::msg::Temperature::ConstSharedPtr message) {
          RCLCPP_INFO(get_logger(), "IMU temperature=%.3f C", message->temperature);
        });
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr imu_subscription_;
  rclcpp::Subscription<sensor_msgs::msg::Temperature>::SharedPtr temperature_subscription_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImuRos2Subscriber>());
  rclcpp::shutdown();
  return 0;
}

File: applications/holoscan_ros2/adi/adi_imu/python/adi_imu_ros2_subscriber.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu, Temperature


class ImuRos2Subscriber(Node):
    def __init__(self):
        super().__init__("adi_imu_ros2_subscriber")
        self.create_subscription(Imu, "imu/data_raw", self.on_imu, 10)
        self.create_subscription(Temperature, "imu/temperature", self.on_temperature, 10)

    def on_imu(self, message):
        self.get_logger().info(
            "IMU gyro=(%.4f, %.4f, %.4f) accel=(%.4f, %.4f, %.4f)"
            % (
                message.angular_velocity.x,
                message.angular_velocity.y,
                message.angular_velocity.z,
                message.linear_acceleration.x,
                message.linear_acceleration.y,
                message.linear_acceleration.z,
            )
        )

    def on_temperature(self, message):
        self.get_logger().info("IMU temperature=%.3f C" % message.temperature)


def main():
    rclpy.init()
    node = ImuRos2Subscriber()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
